adlfs/transfer.py

```diff
@@ -0,0 +1,368 @@
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+"""
+High performance multi-threaded module to up/download
+
+Calls method in `core` with thread pool executor to ensure the network
+is used to its maximum throughput.
+
+Only implements upload and download of (massive) files and directory trees.
+"""
+from collections import namedtuple
+from concurrent.futures import ThreadPoolExecutor
+import logging
+import multiprocessing
+import os
+import pickle
+import threading
+import time
+import uuid
+
+from adlfs.utils import clamp
+
+MAX_THREADS = 64
+
+logger = logging.getLogger(__name__)
+
+File = namedtuple('File', 'src dst state nbytes start stop chunks')
+Chunk = namedtuple('Chunk', 'name state offset retries')
+
+
+class DisjointState(object):
+    """
+    Collection of disjoint sets containing state about objects.
+
+    Internal class used by `ADLTransferClient`.
+    """
+    def __init__(self, *states):
+        self._states = {state: set() for state in states}
+        self._objects = {}
+
+    @property
+    def states(self):
+        return list(self._states)
+
+    @property
+    def objects(self):
+        return list(self._objects)
+
+    def __iter__(self):
+        return iter(self._objects.items())
+
+    def __getitem__(self, obj):
+        return self._objects[obj]
+
+    def __setitem__(self, obj, state):
+        if obj in self._objects:
+            self._states[self._objects[obj]].discard(obj)
+        self._states[state].add(obj)
+        self._objects[obj] = state
+
+    def contains_all(self, state):
+        objs = self._states[state]
+        return len(objs) > 0 and len(self.objects) - len(objs) == 0
+
+    def contains_none(self, *states):
+        return all([len(self._states[state]) == 0 for state in states])
+
+    def __str__(self):
+        status = " ".join(
+            ["%s=%d" % (s, len(self._states[s])) for s in self._states])
+        return "<DisjointState: " + status + ">"
+
+    __repr__ = __str__
+
+
+class ADLTransferClient(object):
+    """
+    Client for transferring data from/to Azure DataLake Store
+
+    Parameters
+    ----------
+    adlfs: ADL filesystem instance
+    name: str
+        Unique ID used for persistence.
+    transfer: callable
+        Function or callable object invoked when transferring chunks. See
+        ``Function Signatures``.
+    merge: callable [None]
+        Function or callable object invoked when merging chunks. If None,
+        then merging is skipped. See ``Function Signatures``.
+    nthreads: int [None]
+        Number of threads to use (range is [1, MAX_THREADS]). If None, uses the
+        number of cores.
+    chunksize: int [2**26]
+        Number of bytes in each chunk for splitting big files. Files smaller
+        than this number will always be transferred in a single thread.
```

Comment:
The best value for downloads (not uploads) was found to be smaller than 256MB (which may impose memory problems), that's where the number came from. Whether the default of the combined transfer class should be that of the upload or download shouldn't matter, since each subclass has its own default too. Maybe there should be no default here at all.

(NB: uploading of files is always in 4MB chunks or smaller, but this logic is in the underlying ADLFile class)